crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -0,0 +1,564 @@
+#[derive(Debug)]
+pub struct PaginationResult {
+    /// Events returned during this pagination.
+    ///
+    /// If this is the result of a backward pagination, then the events are in
+    /// reverse topological order.
+    ///
+    /// If this is the result of a forward pagination, then the events are in
+    /// topological order.
+    pub events: Vec<TimelineEvent>,
+
+    /// Did we hit an end of the timeline?
+    ///
+    /// If this is the result of a backward pagination, this means we hit the
+    /// *start* of the timeline.
+    ///
+    /// If this is the result of a forward pagination, this means we hit the
+    /// *end* of the timeline.
+    pub hit_end_of_timeline: bool,
+}
+
+/// The result of an initial [`Paginator::start_from`] query.
+#[derive(Debug)]
+pub struct StartFromResult {
+    /// All the events returned during this pagination, in topological ordering.
+    pub events: Vec<TimelineEvent>,
+
+    /// Whether the /context query returned a previous batch token.
+    pub has_prev: bool,
+
+    /// Whether the /context query returned a next batch token.
+    pub has_next: bool,
+}
+
+impl Paginator {
+    /// Create a new [`Paginator`], given a room implementation.
+    pub fn new(room: Box<dyn PaginableRoom>) -> Self {
+        Self {
+            room,
+            state: SharedObservable::new(PaginatorState::Initial),
+            prev_batch_token: Mutex::new(None),
+            next_batch_token: Mutex::new(None),
+        }
+    }
+
+    /// Check if the current state of the paginator matches the expected one.
+    fn check_state(&self, expected: PaginatorState) -> Result<(), PaginatorError> {
+        let actual = self.state.get();
+        if actual != expected {
+            return Err(PaginatorError::InvalidPreviousState { expected, actual });
+        }
+        Ok(())
+    }
+
+    /// Returns a subscriber to the internal [`PaginatorState`] machine.
+    pub fn state(&self) -> Subscriber<PaginatorState> {
+        self.state.subscribe()
+    }
+
+    /// Starts the pagination from the initial event.
+    ///
+    /// Only works for fresh [`Paginator`] objects, which are in the
+    /// [`PaginatorState::Initial`] state.
+    pub async fn start_from(&self, event_id: &EventId) -> Result<StartFromResult, PaginatorError> {
+        self.check_state(PaginatorState::Initial)?;
+
+        // Note: it's possible two callers have checked the state and both figured it's
+        // initial. This check makes sure there's at most one which can set the
+        // state to FetchingTargetEvent, preventing a race condition.
+        if self.state.set_if_not_eq(PaginatorState::FetchingTargetEvent).is_none() {
+            return Err(PaginatorError::InvalidPreviousState {
+                expected: PaginatorState::Initial,
+                actual: PaginatorState::FetchingTargetEvent,
+            });
+        }
+
+        // TODO: do we want to lazy load members?
```

Comment:
Members seem irrelevant here, so maybe not?

Reply:
They are relevant! One may click a permalink to a room they've never visited, and as a result have no information whatsoever with respect to the member profiles, so we'd need to process the room member data there. Leaving for later.